### ISSUE TITLE
Update Guava to 21.0

### ISSUE THREAD

**Issue by donat** [eclipse/buildship#285]
<a href="https://github.com/donat"><img src="https://avatars.githubusercontent.com/u/419883?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [donat](https://github.com/donat)**
_Thursday Jan 12, 2017 at 09:27 GMT_

----

## Motivation
The IP team approved Guava 18: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12184
The latest eclipse orbit builds already contain this dependency: http://download.eclipse.org/tools/orbit/downloads/drops/S20161205183421

## Proposed Change
Update guava dependency from 15.0 to 18.0.

## Code review

- [ ] Does the change work?
- [ ] Is the change sufficiently tested?
- [ ] Is the public API modification handled appropriately?
- [ ] Is it sufficiently documented?
- [ ] Is it implemented well?

See also the [full checklist](https://github.com/gradle/process/wiki/Process#implementation-checklist).

**Comment by fbricon** [comment]
Can you consider moving to 21.0, as per https://dev.eclipse.org/mhonarc/lists/cross-project-issues-dev/msg14431.html?

In [eclipse.jdt.ls](https://github.com/eclipse/eclipse.jdt.ls), which consumes BuildShip, since getting the latest bits from m2e depending on Guava 21, we have a new error, because Buildship re-exports Guava as part of its APIs, which leads to:

```
java.lang.LinkageError: loader constraint violation: loader (instance of org/eclipse/osgi/internal/loader/EquinoxClassLoader) previously initiated loading for a different type with name "com/google/common/base/Optional"
	at org.eclipse.jdt.ls.core.internal.managers.GradleBuildSupport.update(GradleBuildSupport.java:43)
```

See https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/GradleBuildSupport.java#L43

**Comment by donat** [comment]
I agree we should upgrade to Guava 21.0. 

There is one catch though: Buildship currently supports Java 1.7 environments, where Guava 21 requires Java 8. We could solve that by using the existing Guava version for Eclipse 4.5 and earlier versions.

**Comment by oehme** [comment]
We could also accept a version range, whatever works best.

**Comment by donat** [comment]
I've added a new build plan to CI to build and create releases against Eclipse Oxygen.

**Comment by donat** [comment]
The snapshot builds for Eclipse 4.7 are available here: http://download.eclipse.org/buildship/updates/e47/snapshots/2.x

**Comment by donat** [comment]
@fbricon to provide the fix for our users we'll have to revert Guava to version 15.0 in the 2.1.2 hotfix release (#529). We'll fix the situation in Buildship 2.2 (#530).

**Comment by UdoW** [comment]
While testing a self-built "buildship 3.x" version with an upgraded Guava version of 21.0.0 I got the error message from the validator of the Eclipse Runtime configuration dialog that these dependencies use old Guava versions, too:

- com.gradleware.tooling.client
- com.gradleware.tooling.model
- com.gradleware.tooling.utils

These have a version constraint on Guava like this: [15.0.0,16.0.0)

**Comment by UdoW** [comment]
For my proof-of-concept I also patched the above 3 bundles to use Guava 21.0.0 and Java 8. 
I got it running:
![image](https://user-images.githubusercontent.com/12225215/44703171-e9417e80-aa96-11e8-9147-8bb24c906a54.png)

**Comment by donat** [comment]
The `com.gradleware.*` plugins have been removed from Buildship 3.0. You may need to uninstall Buildship 2.x before you can install Buildship 3.0. 

We have recently moved Buildship to Java 8, so there are no more blockers for the Guava 15 -> 21 update. In fact, I've already done it (https://github.com/eclipse/buildship/commit/17ade2cfe9731e27ade0fdedd8575c41293a59b4), and the [latest snapshot update site](http://download.eclipse.org/buildship/updates/e48/snapshots/3.x/3.0.0.v20180827-2310-s) contains the change.
